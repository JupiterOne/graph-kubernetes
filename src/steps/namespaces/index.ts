import { IntegrationStep } from '@jupiterone/integration-sdk-core';
import { IntegrationConfig, IntegrationStepContext } from '../../config';
import { Entities, IntegrationSteps } from '../constants';
import { createNamespaceEntity } from './converters';
import getOrCreateAPIClient from '../../kubernetes/getOrCreateAPIClient';

export async function fetchNamespaces(
  context: IntegrationStepContext,
): Promise<void> {
  const { instance, jobState } = context;
  const { config } = instance;

  const client = getOrCreateAPIClient(config);

  await client.iterateNamespaces(async (namespace) => {
    const namespaceEntity = createNamespaceEntity(namespace);
    await jobState.addEntity(namespaceEntity);
  });
}

export const namespaceSteps: IntegrationStep<IntegrationConfig>[] = [
  {
    id: IntegrationSteps.NAMESPACES,
    name: 'Fetch Namespaces',
    entities: [Entities.NAMESPACE],
    relationships: [],
    dependsOn: [],
    executionHandler: fetchNamespaces,
  },
];
