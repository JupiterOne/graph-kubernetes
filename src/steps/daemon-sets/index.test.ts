import { buildContainerSpecDaemonsetRelationship, fetchDaemonSets } from '.';
import { fetchNamespaces } from '../namespaces';
import { createDataCollectionTest } from '../../../test/recording';
import { integrationConfig } from '../../../test/config';
import { Entities, Relationships } from '../constants';
import { RelationshipClass } from '@jupiterone/data-model';

describe('#fetchDaemonSets', () => {
  test('should collect data', async () => {
    await createDataCollectionTest({
      recordingName: 'fetchDaemonSets',
      recordingDirectory: __dirname,
      integrationConfig,
      stepFunctions: [
        fetchNamespaces,
        fetchDaemonSets,
        buildContainerSpecDaemonsetRelationship,
      ],
      entitySchemaMatchers: [
        {
          _type: Entities.CRONJOB._type,
          matcher: {
            _class: ['Deployment'],
            schema: {
              additionalProperties: false,
              properties: {
                _type: { const: 'kube_daemon_set' },
                _rawData: {
                  type: 'array',
                  items: { type: 'object' },
                },
                name: { type: 'string' },
                displayName: { type: 'string' },
                generation: { type: 'number' },
                deletionGracePeriodSeconds: { type: 'string' },
                resourceVersion: { type: 'string' },
                createdOn: { type: 'number' },
                deletedOn: { type: 'number' },
                minReadySeconds: { type: 'number' },
                revisionHistoryLimit: { type: 'number' },
                'strategy.type': { type: 'string' },
                'strategy.maxUnavailable': { type: 'string' },
                'status.collisionCount': { type: 'number' },
                'status.currentNumberScheduled': { type: 'number' },
                'status.desiredNumberScheduled': { type: 'number' },
                'status.numberAvailable': { type: 'number' },
                'status.numberMisscheduled': { type: 'number' },
                'status.numberReady': { type: 'number' },
                'status.numberUnavailable': { type: 'number' },
                'status.observedGeneration': { type: 'number' },
                'status.updatedNumberScheduled': { type: 'number' },
              },
            },
          },
        },
      ],
      relationshipSchemaMatchers: [
        {
          _type: Relationships.NAMESPACE_CONTAINS_DAEMONSET._type,
          matcher: {
            schema: {
              properties: {
                _class: { const: RelationshipClass.CONTAINS },
                _type: { const: 'kube_namespace_contains_daemon_set' },
              },
            },
          },
        },
        {
          _type: Relationships.CONTAINER_SPEC_HAS_DAEMONSET._type,
          matcher: {
            schema: {
              properties: {
                _class: { const: RelationshipClass.HAS },
                _type: { const: 'kube_container_spec_has_daemon_set' },
              },
            },
          },
        },
      ],
    });
  });
});
