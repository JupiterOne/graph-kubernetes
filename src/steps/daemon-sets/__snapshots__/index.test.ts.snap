// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#fetchDaemonSets should collect data: jobState 1`] = `
Object {
  "collectedEntities": Array [
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "9e9e231b-89d5-4ac7-8ec4-855791482c2f",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": V1Namespace {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": undefined,
              "creationTimestamp": 2024-06-19T08:11:16.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": Object {
                "kubernetes.io/metadata.name": "default",
              },
              "managedFields": Array [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": Object {
                    "f:metadata": Object {
                      "f:labels": Object {
                        ".": Object {},
                        "f:kubernetes.io/metadata.name": Object {},
                      },
                    },
                  },
                  "manager": "kube-apiserver",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-06-19T08:11:16.000Z,
                },
              ],
              "name": "default",
              "namespace": undefined,
              "ownerReferences": undefined,
              "resourceVersion": "42",
              "selfLink": undefined,
              "uid": "9e9e231b-89d5-4ac7-8ec4-855791482c2f",
            },
            "spec": V1NamespaceSpec {
              "finalizers": Array [
                "kubernetes",
              ],
            },
            "status": V1NamespaceStatus {
              "conditions": undefined,
              "phase": "Active",
            },
          },
        },
      ],
      "_type": "kube_namespace",
      "active": undefined,
      "createdOn": 1718784676000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "default",
      "finalizers": Array [
        "kubernetes",
      ],
      "generation": undefined,
      "name": "default",
      "resourceVersion": "42",
      "status.phase": "Active",
    },
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "9d2b05a5-c965-4a88-8b01-b9d28aed519a",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": V1Namespace {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": undefined,
              "creationTimestamp": 2024-06-19T08:11:16.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": Object {
                "kubernetes.io/metadata.name": "kube-node-lease",
              },
              "managedFields": Array [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": Object {
                    "f:metadata": Object {
                      "f:labels": Object {
                        ".": Object {},
                        "f:kubernetes.io/metadata.name": Object {},
                      },
                    },
                  },
                  "manager": "kube-apiserver",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-06-19T08:11:16.000Z,
                },
              ],
              "name": "kube-node-lease",
              "namespace": undefined,
              "ownerReferences": undefined,
              "resourceVersion": "38",
              "selfLink": undefined,
              "uid": "9d2b05a5-c965-4a88-8b01-b9d28aed519a",
            },
            "spec": V1NamespaceSpec {
              "finalizers": Array [
                "kubernetes",
              ],
            },
            "status": V1NamespaceStatus {
              "conditions": undefined,
              "phase": "Active",
            },
          },
        },
      ],
      "_type": "kube_namespace",
      "active": undefined,
      "createdOn": 1718784676000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "kube-node-lease",
      "finalizers": Array [
        "kubernetes",
      ],
      "generation": undefined,
      "name": "kube-node-lease",
      "resourceVersion": "38",
      "status.phase": "Active",
    },
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "aee57e5e-0611-4ca8-89f3-64bf6fdaac8c",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": V1Namespace {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": undefined,
              "creationTimestamp": 2024-06-19T08:11:16.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": Object {
                "kubernetes.io/metadata.name": "kube-public",
              },
              "managedFields": Array [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": Object {
                    "f:metadata": Object {
                      "f:labels": Object {
                        ".": Object {},
                        "f:kubernetes.io/metadata.name": Object {},
                      },
                    },
                  },
                  "manager": "kube-apiserver",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-06-19T08:11:16.000Z,
                },
              ],
              "name": "kube-public",
              "namespace": undefined,
              "ownerReferences": undefined,
              "resourceVersion": "26",
              "selfLink": undefined,
              "uid": "aee57e5e-0611-4ca8-89f3-64bf6fdaac8c",
            },
            "spec": V1NamespaceSpec {
              "finalizers": Array [
                "kubernetes",
              ],
            },
            "status": V1NamespaceStatus {
              "conditions": undefined,
              "phase": "Active",
            },
          },
        },
      ],
      "_type": "kube_namespace",
      "active": undefined,
      "createdOn": 1718784676000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "kube-public",
      "finalizers": Array [
        "kubernetes",
      ],
      "generation": undefined,
      "name": "kube-public",
      "resourceVersion": "26",
      "status.phase": "Active",
    },
    Object {
      "_class": Array [
        "Group",
      ],
      "_key": "ced267a3-528f-46b5-a289-cbf7068d8fc9",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": V1Namespace {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": undefined,
              "creationTimestamp": 2024-06-19T08:11:16.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": undefined,
              "labels": Object {
                "kubernetes.io/metadata.name": "kube-system",
              },
              "managedFields": Array [
                V1ManagedFieldsEntry {
                  "apiVersion": "v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": Object {
                    "f:metadata": Object {
                      "f:labels": Object {
                        ".": Object {},
                        "f:kubernetes.io/metadata.name": Object {},
                      },
                    },
                  },
                  "manager": "kube-apiserver",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-06-19T08:11:16.000Z,
                },
              ],
              "name": "kube-system",
              "namespace": undefined,
              "ownerReferences": undefined,
              "resourceVersion": "12",
              "selfLink": undefined,
              "uid": "ced267a3-528f-46b5-a289-cbf7068d8fc9",
            },
            "spec": V1NamespaceSpec {
              "finalizers": Array [
                "kubernetes",
              ],
            },
            "status": V1NamespaceStatus {
              "conditions": undefined,
              "phase": "Active",
            },
          },
        },
      ],
      "_type": "kube_namespace",
      "active": undefined,
      "createdOn": 1718784676000,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "kube-system",
      "finalizers": Array [
        "kubernetes",
      ],
      "generation": undefined,
      "name": "kube-system",
      "resourceVersion": "12",
      "status.phase": "Active",
    },
    Object {
      "_class": Array [
        "Deployment",
      ],
      "_key": "4b30d371-4797-45c6-82e5-c7c8d8e5ca06",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": V1DaemonSet {
            "apiVersion": undefined,
            "kind": undefined,
            "metadata": V1ObjectMeta {
              "annotations": Object {
                "deprecated.daemonset.template.generation": "1",
              },
              "creationTimestamp": 2024-06-19T08:11:19.000Z,
              "deletionGracePeriodSeconds": undefined,
              "deletionTimestamp": undefined,
              "finalizers": undefined,
              "generateName": undefined,
              "generation": 1,
              "labels": Object {
                "k8s-app": "kube-proxy",
              },
              "managedFields": Array [
                V1ManagedFieldsEntry {
                  "apiVersion": "apps/v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": Object {
                    "f:metadata": Object {
                      "f:annotations": Object {
                        ".": Object {},
                        "f:deprecated.daemonset.template.generation": Object {},
                      },
                      "f:labels": Object {
                        ".": Object {},
                        "f:k8s-app": Object {},
                      },
                    },
                    "f:spec": Object {
                      "f:revisionHistoryLimit": Object {},
                      "f:selector": Object {},
                      "f:template": Object {
                        "f:metadata": Object {
                          "f:labels": Object {
                            ".": Object {},
                            "f:k8s-app": Object {},
                          },
                        },
                        "f:spec": Object {
                          "f:containers": Object {
                            "k:{\\"name\\":\\"kube-proxy\\"}": Object {
                              ".": Object {},
                              "f:command": Object {},
                              "f:env": Object {
                                ".": Object {},
                                "k:{\\"name\\":\\"NODE_NAME\\"}": Object {
                                  ".": Object {},
                                  "f:name": Object {},
                                  "f:valueFrom": Object {
                                    ".": Object {},
                                    "f:fieldRef": Object {},
                                  },
                                },
                              },
                              "f:image": Object {},
                              "f:imagePullPolicy": Object {},
                              "f:name": Object {},
                              "f:resources": Object {},
                              "f:securityContext": Object {
                                ".": Object {},
                                "f:privileged": Object {},
                              },
                              "f:terminationMessagePath": Object {},
                              "f:terminationMessagePolicy": Object {},
                              "f:volumeMounts": Object {
                                ".": Object {},
                                "k:{\\"mountPath\\":\\"/lib/modules\\"}": Object {
                                  ".": Object {},
                                  "f:mountPath": Object {},
                                  "f:name": Object {},
                                  "f:readOnly": Object {},
                                },
                                "k:{\\"mountPath\\":\\"/run/xtables.lock\\"}": Object {
                                  ".": Object {},
                                  "f:mountPath": Object {},
                                  "f:name": Object {},
                                },
                                "k:{\\"mountPath\\":\\"/var/lib/kube-proxy\\"}": Object {
                                  ".": Object {},
                                  "f:mountPath": Object {},
                                  "f:name": Object {},
                                },
                              },
                            },
                          },
                          "f:dnsPolicy": Object {},
                          "f:hostNetwork": Object {},
                          "f:nodeSelector": Object {},
                          "f:priorityClassName": Object {},
                          "f:restartPolicy": Object {},
                          "f:schedulerName": Object {},
                          "f:securityContext": Object {},
                          "f:serviceAccount": Object {},
                          "f:serviceAccountName": Object {},
                          "f:terminationGracePeriodSeconds": Object {},
                          "f:tolerations": Object {},
                          "f:volumes": Object {
                            ".": Object {},
                            "k:{\\"name\\":\\"kube-proxy\\"}": Object {
                              ".": Object {},
                              "f:configMap": Object {
                                ".": Object {},
                                "f:defaultMode": Object {},
                                "f:name": Object {},
                              },
                              "f:name": Object {},
                            },
                            "k:{\\"name\\":\\"lib-modules\\"}": Object {
                              ".": Object {},
                              "f:hostPath": Object {
                                ".": Object {},
                                "f:path": Object {},
                                "f:type": Object {},
                              },
                              "f:name": Object {},
                            },
                            "k:{\\"name\\":\\"xtables-lock\\"}": Object {
                              ".": Object {},
                              "f:hostPath": Object {
                                ".": Object {},
                                "f:path": Object {},
                                "f:type": Object {},
                              },
                              "f:name": Object {},
                            },
                          },
                        },
                      },
                      "f:updateStrategy": Object {
                        "f:rollingUpdate": Object {
                          ".": Object {},
                          "f:maxSurge": Object {},
                          "f:maxUnavailable": Object {},
                        },
                        "f:type": Object {},
                      },
                    },
                  },
                  "manager": "kubeadm",
                  "operation": "Update",
                  "subresource": undefined,
                  "time": 2024-06-19T08:11:19.000Z,
                },
                V1ManagedFieldsEntry {
                  "apiVersion": "apps/v1",
                  "fieldsType": "FieldsV1",
                  "fieldsV1": Object {
                    "f:status": Object {
                      "f:currentNumberScheduled": Object {},
                      "f:desiredNumberScheduled": Object {},
                      "f:numberAvailable": Object {},
                      "f:numberReady": Object {},
                      "f:observedGeneration": Object {},
                      "f:updatedNumberScheduled": Object {},
                    },
                  },
                  "manager": "kube-controller-manager",
                  "operation": "Update",
                  "subresource": "status",
                  "time": 2024-06-19T08:11:34.000Z,
                },
              ],
              "name": "kube-proxy",
              "namespace": "kube-system",
              "ownerReferences": undefined,
              "resourceVersion": "406",
              "selfLink": undefined,
              "uid": "4b30d371-4797-45c6-82e5-c7c8d8e5ca06",
            },
            "spec": V1DaemonSetSpec {
              "minReadySeconds": undefined,
              "revisionHistoryLimit": 10,
              "selector": V1LabelSelector {
                "matchExpressions": undefined,
                "matchLabels": Object {
                  "k8s-app": "kube-proxy",
                },
              },
              "template": V1PodTemplateSpec {
                "metadata": V1ObjectMeta {
                  "annotations": undefined,
                  "creationTimestamp": null,
                  "deletionGracePeriodSeconds": undefined,
                  "deletionTimestamp": undefined,
                  "finalizers": undefined,
                  "generateName": undefined,
                  "generation": undefined,
                  "labels": Object {
                    "k8s-app": "kube-proxy",
                  },
                  "managedFields": undefined,
                  "name": undefined,
                  "namespace": undefined,
                  "ownerReferences": undefined,
                  "resourceVersion": undefined,
                  "selfLink": undefined,
                  "uid": undefined,
                },
                "spec": V1PodSpec {
                  "activeDeadlineSeconds": undefined,
                  "affinity": undefined,
                  "automountServiceAccountToken": undefined,
                  "containers": Array [
                    V1Container {
                      "args": undefined,
                      "command": Array [
                        "/usr/local/bin/kube-proxy",
                        "--config=/var/lib/kube-proxy/config.conf",
                        "--hostname-override=$(NODE_NAME)",
                      ],
                      "env": Array [
                        V1EnvVar {
                          "name": "NODE_NAME",
                          "value": undefined,
                          "valueFrom": V1EnvVarSource {
                            "configMapKeyRef": undefined,
                            "fieldRef": V1ObjectFieldSelector {
                              "apiVersion": "v1",
                              "fieldPath": "spec.nodeName",
                            },
                            "resourceFieldRef": undefined,
                            "secretKeyRef": undefined,
                          },
                        },
                      ],
                      "envFrom": undefined,
                      "image": "registry.k8s.io/kube-proxy:v1.30.0",
                      "imagePullPolicy": "IfNotPresent",
                      "lifecycle": undefined,
                      "livenessProbe": undefined,
                      "name": "kube-proxy",
                      "ports": undefined,
                      "readinessProbe": undefined,
                      "resizePolicy": undefined,
                      "resources": V1ResourceRequirements {
                        "claims": undefined,
                        "limits": undefined,
                        "requests": undefined,
                      },
                      "restartPolicy": undefined,
                      "securityContext": V1SecurityContext {
                        "allowPrivilegeEscalation": undefined,
                        "capabilities": undefined,
                        "privileged": true,
                        "procMount": undefined,
                        "readOnlyRootFilesystem": undefined,
                        "runAsGroup": undefined,
                        "runAsNonRoot": undefined,
                        "runAsUser": undefined,
                        "seLinuxOptions": undefined,
                        "seccompProfile": undefined,
                        "windowsOptions": undefined,
                      },
                      "startupProbe": undefined,
                      "stdin": undefined,
                      "stdinOnce": undefined,
                      "terminationMessagePath": "/dev/termination-log",
                      "terminationMessagePolicy": "File",
                      "tty": undefined,
                      "volumeDevices": undefined,
                      "volumeMounts": Array [
                        V1VolumeMount {
                          "mountPath": "/var/lib/kube-proxy",
                          "mountPropagation": undefined,
                          "name": "kube-proxy",
                          "readOnly": undefined,
                          "subPath": undefined,
                          "subPathExpr": undefined,
                        },
                        V1VolumeMount {
                          "mountPath": "/run/xtables.lock",
                          "mountPropagation": undefined,
                          "name": "xtables-lock",
                          "readOnly": undefined,
                          "subPath": undefined,
                          "subPathExpr": undefined,
                        },
                        V1VolumeMount {
                          "mountPath": "/lib/modules",
                          "mountPropagation": undefined,
                          "name": "lib-modules",
                          "readOnly": true,
                          "subPath": undefined,
                          "subPathExpr": undefined,
                        },
                      ],
                      "workingDir": undefined,
                    },
                  ],
                  "dnsConfig": undefined,
                  "dnsPolicy": "ClusterFirst",
                  "enableServiceLinks": undefined,
                  "ephemeralContainers": undefined,
                  "hostAliases": undefined,
                  "hostIPC": undefined,
                  "hostNetwork": true,
                  "hostPID": undefined,
                  "hostUsers": undefined,
                  "hostname": undefined,
                  "imagePullSecrets": undefined,
                  "initContainers": undefined,
                  "nodeName": undefined,
                  "nodeSelector": Object {
                    "kubernetes.io/os": "linux",
                  },
                  "os": undefined,
                  "overhead": undefined,
                  "preemptionPolicy": undefined,
                  "priority": undefined,
                  "priorityClassName": "system-node-critical",
                  "readinessGates": undefined,
                  "resourceClaims": undefined,
                  "restartPolicy": "Always",
                  "runtimeClassName": undefined,
                  "schedulerName": "default-scheduler",
                  "schedulingGates": undefined,
                  "securityContext": V1PodSecurityContext {
                    "fsGroup": undefined,
                    "fsGroupChangePolicy": undefined,
                    "runAsGroup": undefined,
                    "runAsNonRoot": undefined,
                    "runAsUser": undefined,
                    "seLinuxOptions": undefined,
                    "seccompProfile": undefined,
                    "supplementalGroups": undefined,
                    "sysctls": undefined,
                    "windowsOptions": undefined,
                  },
                  "serviceAccount": "kube-proxy",
                  "serviceAccountName": "kube-proxy",
                  "setHostnameAsFQDN": undefined,
                  "shareProcessNamespace": undefined,
                  "subdomain": undefined,
                  "terminationGracePeriodSeconds": 30,
                  "tolerations": Array [
                    V1Toleration {
                      "effect": undefined,
                      "key": undefined,
                      "operator": "Exists",
                      "tolerationSeconds": undefined,
                      "value": undefined,
                    },
                  ],
                  "topologySpreadConstraints": undefined,
                  "volumes": Array [
                    V1Volume {
                      "awsElasticBlockStore": undefined,
                      "azureDisk": undefined,
                      "azureFile": undefined,
                      "cephfs": undefined,
                      "cinder": undefined,
                      "configMap": V1ConfigMapVolumeSource {
                        "defaultMode": 420,
                        "items": undefined,
                        "name": "kube-proxy",
                        "optional": undefined,
                      },
                      "csi": undefined,
                      "downwardAPI": undefined,
                      "emptyDir": undefined,
                      "ephemeral": undefined,
                      "fc": undefined,
                      "flexVolume": undefined,
                      "flocker": undefined,
                      "gcePersistentDisk": undefined,
                      "gitRepo": undefined,
                      "glusterfs": undefined,
                      "hostPath": undefined,
                      "iscsi": undefined,
                      "name": "kube-proxy",
                      "nfs": undefined,
                      "persistentVolumeClaim": undefined,
                      "photonPersistentDisk": undefined,
                      "portworxVolume": undefined,
                      "projected": undefined,
                      "quobyte": undefined,
                      "rbd": undefined,
                      "scaleIO": undefined,
                      "secret": undefined,
                      "storageos": undefined,
                      "vsphereVolume": undefined,
                    },
                    V1Volume {
                      "awsElasticBlockStore": undefined,
                      "azureDisk": undefined,
                      "azureFile": undefined,
                      "cephfs": undefined,
                      "cinder": undefined,
                      "configMap": undefined,
                      "csi": undefined,
                      "downwardAPI": undefined,
                      "emptyDir": undefined,
                      "ephemeral": undefined,
                      "fc": undefined,
                      "flexVolume": undefined,
                      "flocker": undefined,
                      "gcePersistentDisk": undefined,
                      "gitRepo": undefined,
                      "glusterfs": undefined,
                      "hostPath": V1HostPathVolumeSource {
                        "path": "/run/xtables.lock",
                        "type": "FileOrCreate",
                      },
                      "iscsi": undefined,
                      "name": "xtables-lock",
                      "nfs": undefined,
                      "persistentVolumeClaim": undefined,
                      "photonPersistentDisk": undefined,
                      "portworxVolume": undefined,
                      "projected": undefined,
                      "quobyte": undefined,
                      "rbd": undefined,
                      "scaleIO": undefined,
                      "secret": undefined,
                      "storageos": undefined,
                      "vsphereVolume": undefined,
                    },
                    V1Volume {
                      "awsElasticBlockStore": undefined,
                      "azureDisk": undefined,
                      "azureFile": undefined,
                      "cephfs": undefined,
                      "cinder": undefined,
                      "configMap": undefined,
                      "csi": undefined,
                      "downwardAPI": undefined,
                      "emptyDir": undefined,
                      "ephemeral": undefined,
                      "fc": undefined,
                      "flexVolume": undefined,
                      "flocker": undefined,
                      "gcePersistentDisk": undefined,
                      "gitRepo": undefined,
                      "glusterfs": undefined,
                      "hostPath": V1HostPathVolumeSource {
                        "path": "/lib/modules",
                        "type": "",
                      },
                      "iscsi": undefined,
                      "name": "lib-modules",
                      "nfs": undefined,
                      "persistentVolumeClaim": undefined,
                      "photonPersistentDisk": undefined,
                      "portworxVolume": undefined,
                      "projected": undefined,
                      "quobyte": undefined,
                      "rbd": undefined,
                      "scaleIO": undefined,
                      "secret": undefined,
                      "storageos": undefined,
                      "vsphereVolume": undefined,
                    },
                  ],
                },
              },
              "updateStrategy": V1DaemonSetUpdateStrategy {
                "rollingUpdate": V1RollingUpdateDaemonSet {
                  "maxSurge": 0,
                  "maxUnavailable": 1,
                },
                "type": "RollingUpdate",
              },
            },
            "status": V1DaemonSetStatus {
              "collisionCount": undefined,
              "conditions": undefined,
              "currentNumberScheduled": 1,
              "desiredNumberScheduled": 1,
              "numberAvailable": 1,
              "numberMisscheduled": 0,
              "numberReady": 1,
              "numberUnavailable": undefined,
              "observedGeneration": 1,
              "updatedNumberScheduled": 1,
            },
          },
        },
      ],
      "_type": "kube_daemon_set",
      "active": undefined,
      "createdOn": 1718784679000,
      "deletedOn": undefined,
      "deletionGracePeriodSeconds": undefined,
      "displayName": "kube-proxy",
      "generation": 1,
      "minReadySeconds": undefined,
      "name": "kube-proxy",
      "resourceVersion": "406",
      "revisionHistoryLimit": 10,
      "status.collisionCount": undefined,
      "status.currentNumberScheduled": 1,
      "status.desiredNumberScheduled": 1,
      "status.numberAvailable": 1,
      "status.numberMisscheduled": 0,
      "status.numberReady": 1,
      "status.numberUnavailable": undefined,
      "status.observedGeneration": 1,
      "status.updatedNumberScheduled": 1,
      "strategy.maxUnavailable": 1,
      "strategy.type": "RollingUpdate",
    },
    Object {
      "_class": Array [
        "Configuration",
      ],
      "_key": "daemonset/kube-proxy",
      "_rawData": Array [
        Object {
          "name": "default",
          "rawData": V1Container {
            "args": undefined,
            "command": Array [
              "/usr/local/bin/kube-proxy",
              "--config=/var/lib/kube-proxy/config.conf",
              "--hostname-override=$(NODE_NAME)",
            ],
            "env": Array [
              V1EnvVar {
                "name": "NODE_NAME",
                "value": undefined,
                "valueFrom": V1EnvVarSource {
                  "configMapKeyRef": undefined,
                  "fieldRef": V1ObjectFieldSelector {
                    "apiVersion": "v1",
                    "fieldPath": "spec.nodeName",
                  },
                  "resourceFieldRef": undefined,
                  "secretKeyRef": undefined,
                },
              },
            ],
            "envFrom": undefined,
            "image": "registry.k8s.io/kube-proxy:v1.30.0",
            "imagePullPolicy": "IfNotPresent",
            "lifecycle": undefined,
            "livenessProbe": undefined,
            "name": "kube-proxy",
            "ports": undefined,
            "readinessProbe": undefined,
            "resizePolicy": undefined,
            "resources": V1ResourceRequirements {
              "claims": undefined,
              "limits": undefined,
              "requests": undefined,
            },
            "restartPolicy": undefined,
            "securityContext": V1SecurityContext {
              "allowPrivilegeEscalation": undefined,
              "capabilities": undefined,
              "privileged": true,
              "procMount": undefined,
              "readOnlyRootFilesystem": undefined,
              "runAsGroup": undefined,
              "runAsNonRoot": undefined,
              "runAsUser": undefined,
              "seLinuxOptions": undefined,
              "seccompProfile": undefined,
              "windowsOptions": undefined,
            },
            "startupProbe": undefined,
            "stdin": undefined,
            "stdinOnce": undefined,
            "terminationMessagePath": "/dev/termination-log",
            "terminationMessagePolicy": "File",
            "tty": undefined,
            "volumeDevices": undefined,
            "volumeMounts": Array [
              V1VolumeMount {
                "mountPath": "/var/lib/kube-proxy",
                "mountPropagation": undefined,
                "name": "kube-proxy",
                "readOnly": undefined,
                "subPath": undefined,
                "subPathExpr": undefined,
              },
              V1VolumeMount {
                "mountPath": "/run/xtables.lock",
                "mountPropagation": undefined,
                "name": "xtables-lock",
                "readOnly": undefined,
                "subPath": undefined,
                "subPathExpr": undefined,
              },
              V1VolumeMount {
                "mountPath": "/lib/modules",
                "mountPropagation": undefined,
                "name": "lib-modules",
                "readOnly": true,
                "subPath": undefined,
                "subPathExpr": undefined,
              },
            ],
            "workingDir": undefined,
          },
        },
      ],
      "_type": "kube_container_spec",
      "createdOn": undefined,
      "displayName": "kube-proxy",
      "image": "registry.k8s.io/kube-proxy:v1.30.0",
      "imagePullPolicy": "IfNotPresent",
      "name": "kube-proxy",
      "securityContext.allowPrivilegeEscalation": undefined,
      "securityContext.privileged": true,
      "securityContext.procMount": undefined,
      "securityContext.readOnlyRootFilesystem": undefined,
      "securityContext.runAsGroup": undefined,
      "securityContext.runAsNonRoot": undefined,
      "securityContext.runAsUser": undefined,
    },
  ],
  "collectedRelationships": Array [
    Object {
      "_class": "CONTAINS",
      "_fromEntityKey": "ced267a3-528f-46b5-a289-cbf7068d8fc9",
      "_key": "ced267a3-528f-46b5-a289-cbf7068d8fc9|contains|4b30d371-4797-45c6-82e5-c7c8d8e5ca06",
      "_toEntityKey": "4b30d371-4797-45c6-82e5-c7c8d8e5ca06",
      "_type": "kube_namespace_contains_daemon_set",
      "displayName": "CONTAINS",
    },
    Object {
      "_class": "HAS",
      "_fromEntityKey": "daemonset/kube-proxy",
      "_key": "daemonset/kube-proxy|has|4b30d371-4797-45c6-82e5-c7c8d8e5ca06",
      "_toEntityKey": "4b30d371-4797-45c6-82e5-c7c8d8e5ca06",
      "_type": "kube_container_spec_has_daemon_set",
      "displayName": "HAS",
    },
  ],
  "encounteredTypes": Array [
    "kube_namespace",
    "kube_daemon_set",
    "kube_namespace_contains_daemon_set",
    "kube_container_spec",
    "kube_container_spec_has_daemon_set",
  ],
  "numCollectedEntities": 6,
  "numCollectedRelationships": 2,
}
`;
